rust/src/builders/tx_builder.rs

```diff
@@ -509,89 +399,33 @@ impl TransactionBuilder {
-    }
-    pub fn add_script_input(&mut self, hash: &ScriptHash, input: &TransactionInput, amount: &Value) {
-        self.inputs.push(TxBuilderInput {
-            input: input.clone(),
-            amount: amount.clone(),
-        });
-        self.input_types.scripts.insert(hash.clone());
-    }
-    pub fn add_bootstrap_input(&mut self, hash: &ByronAddress, input: &TransactionInput, amount: &Value) {
+    pub fn add_input(&mut self, result: &InputBuilderResult) {
         self.inputs.push(TxBuilderInput {
-            input: input.clone(),
-            amount: amount.clone(),
+            input: result.input.clone(),
+            amount: result.utxo_info.amount.clone(),
         });
-        self.input_types.bootstraps.insert(hash.to_bytes());
+        if let Some(ref data) = result.aggregate_witness {
+            self.witness_set_builder.add_input_aggregate_witness_data(data)
+        }
+        self.witness_set_builder.add_required_wits(&result.required_wits);
     }
 
-    pub fn add_input(&mut self, address: &Address, input: &TransactionInput, amount: &Value) {
-        match &BaseAddress::from_address(address) {
-            Some(addr) => {
-                match &addr.payment_cred().to_keyhash() {
-                    Some(hash) => return self.add_key_input(hash, input, amount),
-                    None => (),
-                }
-                match &addr.payment_cred().to_scripthash() {
-                    Some(hash) => return self.add_script_input(hash, input, amount),
-                    None => (),
-                }
-            },
-            None => (),
-        }
-        match &EnterpriseAddress::from_address(address) {
-            Some(addr) => {
-                match &addr.payment_cred().to_keyhash() {
-                    Some(hash) => return self.add_key_input(hash, input, amount),
-                    None => (),
-                }
-                match &addr.payment_cred().to_scripthash() {
-                    Some(hash) => return self.add_script_input(hash, input, amount),
-                    None => (),
-                }
-            },
-            None => (),
-        }
-        match &PointerAddress::from_address(address) {
-            Some(addr) => {
-                match &addr.payment_cred().to_keyhash() {
-                    Some(hash) => return self.add_key_input(hash, input, amount),
-                    None => (),
-                }
-                match &addr.payment_cred().to_scripthash() {
-                    Some(hash) => return self.add_script_input(hash, input, amount),
-                    None => (),
-                }
-            },
-            None => (),
-        }
-        match &ByronAddress::from_address(address) {
-            Some(addr) => self.add_bootstrap_input(addr, input, amount),
-            None => (),
-        }
+    pub fn add_utxo(&mut self, result: &InputBuilderResult) {
+        self.utxos.push(result.clone());
```

Comment:
It feels kind of awkward that this and `add_input` are separate functions. Probably this awkwardness comes from the fact the input selection is too tightly coupled with the tx builder, but probably this should be separated in a follow-up PR. If you really want to include this function, maybe add a warning / reminder that we want to delete this at some point.

Reply:
Yeah, actually I'm going to move CIP2 out or implement [this](https://github.com/cardano-foundation/CIPs/issues/232) instead after this PR is merged.